### PR TITLE
Fix for ObjectDictionaryConverter

### PR DESCRIPTION
**Pull request by sungam3r**
@benmccallum It is for your graphql-dotnet/server#319

### CODE REVIEW

File: src/GraphQL.SystemTextJson/ObjectDictionaryConverterOld.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace GraphQL.SystemTextJson
+{
+    /// <summary>
+    /// A custom JsonConverter for reading a dictionary of objects of their real underlying type.
+    /// </summary>
+    /// <remarks>
+    /// Based on @pekkah's from tanka-graphql.
+    /// </remarks>
+    internal class ObjectDictionaryConverterOld : JsonConverter<Dictionary<string, object>>
```
Comment (sungam3r) [inline src/GraphQL.SystemTextJson/ObjectDictionaryConverterOld.cs]: Will be removed, for demonstration purpose only.

**Comment by sungam3r** [comment]
ApiApprovalTests.PublicApi.GraphQL.approved.txt changed a lot because of sorting feature, `BoolBox` was only added.

**Comment by pekkah** [comment]
There are some tests here you might want to copy https://github.com/pekkah/tanka-graphql/blob/master/tests/graphql.server.links.tests/DTOs/ObjectDictionaryConverterFacts.cs. I hit some issues when using utf8reader without json document ParseValue. Those tests should cover it.

**Comment by sungam3r** [comment]
@pekkah Thanks. I was going to add tests, your code saved me time.

**Comment by sungam3r** [comment]
@benmccallum  I think it is a success.

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.14393.3384 (1607/AnniversaryUpdate/Redstone1)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
Frequency=3515629 Hz, Resolution=284.4441 ns, Timer=TSC
.NET Core SDK=3.1.100
  [Host]     : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT
  DefaultJob : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT


```
|         Method |          Code |         Mean |       Error |      StdDev | Ratio |   Gen 0 |   Gen 1 | Gen 2 | Allocated |
|--------------- |-------------- |-------------:|------------:|------------:|------:|--------:|--------:|------:|----------:|
| **NewtonsoftJson** |         **Empty** |     **594.8 ns** |     **3.26 ns** |     **2.89 ns** |  **1.00** |  **0.8030** |       **-** |     **-** |    **3360 B** |
| SystemTextJson |         Empty |     269.4 ns |     2.48 ns |     2.19 ns |  0.45 |  0.0167 |       - |     - |      72 B |
|                |               |              |             |             |       |         |         |       |           |
| **NewtonsoftJson** | **Introspection** | **622,776.2 ns** | **4,424.20 ns** | **3,921.93 ns** |  **1.00** | **90.8203** | **33.2031** |     **-** |  **470321 B** |
| SystemTextJson | Introspection | 213,722.0 ns | 1,136.23 ns | 1,007.23 ns |  0.34 | 31.4941 | 10.4980 |     - |  138360 B |
|                |               |              |             |             |       |         |         |       |           |
| **NewtonsoftJson** |         **Short** |   **7,967.9 ns** |   **103.90 ns** |    **92.10 ns** |  **1.00** |  **2.3651** |       **-** |     **-** |    **9904 B** |
| SystemTextJson |         Short |   2,596.5 ns |    14.24 ns |    11.12 ns |  0.33 |  0.4692 |       - |     - |    1968 B |

File: src/GraphQL.SystemTextJson/ExecutionResultJsonConverter.cs [hunk 2]
```diff
@@ -11,16 +11,16 @@ public override void Write(Utf8JsonWriter writer, ExecutionResult value, JsonSer
         {
             writer.WriteStartObject();
 
-            // Important: Don't pass the same options down when recursively calling Serialize.
+            // Important: Be careful with passing the same options down when recursively calling Serialize.
```
Comment (sungam3r) [inline src/GraphQL.SystemTextJson/ExecutionResultJsonConverter.cs]: @benmccallum In fact, there are no problems with passing down options if we serialize **another type** of object, no recursion.

File: src/GraphQL.SystemTextJson/ObjectDictionaryConverterOld.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL.SystemTextJson/ObjectDictionaryConverterOld.cs]: removed

**Comment by sungam3r** [comment]
I would like to merge it ASAP because our first implementation is frankly terrible.

File: src/GraphQL.SystemTextJson/ExecutionResultJsonConverter.cs (hunk 2 above)
Comment (benmccallum) [reply src/GraphQL.SystemTextJson/ExecutionResultJsonConverter.cs]: Good to see, that makes me feel way more comfortable that the options will be used in the lower parts.

File: src/GraphQL.SystemTextJson/ObjectDictionaryConverter.cs [hunk 3]
```diff
@@ -1,82 +1,126 @@
...
-            return ReadDictionary(doc.RootElement);
-        }
-
         public override void Write(Utf8JsonWriter writer, Dictionary<string, object> value, JsonSerializerOptions options)
             => throw new NotImplementedException(
                 "This converter currently is only intended to be used to read a JSON object into a strongly-typed representation.");
 
-        private Dictionary<string, object> ReadDictionary(JsonElement element)
+        public override Dictionary<string, object> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            => ReadDictionary(ref reader);
```
Comment (benmccallum) [inline src/GraphQL.SystemTextJson/ObjectDictionaryConverter.cs]: Ahh, I see. Makes sense.

**Comment by benmccallum** [comment]
Yea no worries on the merge. Sorry I've been busy. Everything looks great here. Now the benchmark numbers are impressive! Nice one! 🎉

**Comment by benmccallum** [comment]
You'll definitely want to copy that change into tanka @pekkah. Makes a huge difference!

File: src/GraphQL.SystemTextJson/ObjectDictionaryConverter.cs [hunk 4]
```diff
@@ -1,82 +1,126 @@
...
+                        return bi;
+                }
                 return d;
-            else if (value.TryGetDecimal(out var dd))
-                return dd;
+            }
+            else if (reader.TryGetDecimal(out decimal dm))
+            {
+                // the idea is to see if there is a loss of accuracy of value
+                if (TryGetBigInteger(ref reader, out var bi))
```
Comment (pekkah) [inline src/GraphQL.SystemTextJson/ObjectDictionaryConverter.cs]: You could query if there's a converter for Big Integer from options(GetConverter) and skip the big integer if no converter present.

Comment (sungam3r) [reply src/GraphQL.SystemTextJson/ObjectDictionaryConverter.cs]: Ok, but I do not want to skip.

Comment (sungam3r) [reply src/GraphQL.SystemTextJson/ObjectDictionaryConverter.cs]: But at least I saw that I wrote 2 identical pieces of code.

**Pull request merged**